tools/loader/ImageClassifier.cpp

```diff
@@ -457,34 +492,87 @@ int main(int argc, char **argv) {
+      SMT = bindings->get(outputPH);
     }
     assert(inputImagePH && SMT && "Input and output must be valid.");
-    GLOW_ASSERT(inputImagePH->dims() == inputImageData.dims() &&
+    GLOW_ASSERT(inputImagePH->dims() == inputImageData->dims() &&
                 "New input shape does not match the compiled function.");
 
     // Convert the raw input to fp16. This must be done every time we get new
     // image data.
     if (convertInAndOutToFp16) {
-      inputImageData.convertToType(ElemKind::Float16Ty);
+      inputImageData->convertToType(ElemKind::Float16Ty);
     }
 
     // About to run inference, so update the input image Placeholder's backing
     // Tensor with inputImageData.
-    updateInputPlaceholders(bindings, {inputImagePH}, {&inputImageData});
+    updateInputPlaceholders(*bindings, {inputImagePH}, {inputImageData.get()});
 
     // Perform the inference execution, updating SMT.
-    auto batchSize = inputImageData.dims()[0];
-    loader.runInference(bindings, batchSize);
-
-    // Print the top-k results from the output Softmax tensor.
-    numErrors += processAndPrintResults(SMT, inputImageBatchFilenames);
+    auto batchSize = inputImageData->dims()[0];
+    if (iterationsOpt) {
+      unsigned requestCount = 1;
+      if (miniBatch) {
+        requestCount = iterationsOpt / miniBatch;
+      }
+      // Setup list of inference requests to be run.
+      for (unsigned i = 0; i < requestCount; i++) {
+        auto newImage = llvm::make_unique<Tensor>(inputImageData->getType());
+        auto newContext = llvm::make_unique<ExecutionContext>();
+        auto ph = newContext->getPlaceholderBindings();
+        ph->allocate(inputImagePH);
+        ph->allocate(outputPH);
+        updateInputPlaceholders(*ph, {inputImagePH}, {newImage.get()});
+
+        preLoadedImages.push_back(std::move(newImage));
+        contexts.push_back(std::move(newContext));
+      }
+    } else {
+      loader.runInference(*bindings, batchSize);
+      // Print the top-k results from the output Softmax tensor.
+      numErrors += processAndPrintResults(SMT, inputImageBatchFilenames);
+    }
+  }
+  if (iterationsOpt) {
+    runtime::HostManager *hostManager = loader.getHostManager();
+    std::string name = loader.getFunctionName();
+    std::atomic<unsigned> inFlight(0);
+    std::promise<void> runPromise;
+    auto fut = runPromise.get_future();
+    if (timeOpt) {
+      timer.startTimer();
+    }
+    for (auto &batch : contexts) {
+      llvm::Error runErr = llvm::Error::success();
+      inFlight++;
+      hostManager->runNetwork(
+          name, std::move(batch),
+          [&runErr, &runPromise,
+           &inFlight](runtime::RunIdentifierTy, llvm::Error err,
+                      std::unique_ptr<ExecutionContext> contextPtr) {
+            runErr = std::move(err);
+            if (--inFlight == 0) {
+              runPromise.set_value();
+            }
+          });
+
+      EXIT_ON_ERR(std::move(runErr));
```

Comment:
`runErr` is likely not valid at this point since it's only set after the future is complete. I think for this design where the future is set after `inFlight` results you need to do the equivalent of logical OR on the runErr and check it at the end. Alternatively just exit (crash) in the runNetwork callback.